### PR TITLE
Main with deep

### PR DESCRIPTION
**Pull request by adischw**
merge the deep branch into master, now tests/build should pass

### CODE REVIEW

File: src/block_helpers.rs [hunk 1]
```diff
@@ -1,59 +1,82 @@
...
+}
+
 #[derive(Clone, Debug)]
 #[repr(C)]
 pub struct WeightAndOptimizerData<L:OptimizerTrait> {
     pub weight: f32, 
     pub optimizer_data: L::PerWeightStore,
 }
 
+
```
Comment (SkBlaz) [inline src/block_helpers.rs:34]: It appears [](https://docs.rs/assert_float_eq/latest/assert_float_eq/) crate is used for approximate assertions, might be useful down the road

File: src/block_helpers.rs [hunk 2]
```diff
@@ -88,39 +111,81 @@ pub fn read_weights_only_from_buf2<L:OptimizerTrait>(weights_len: usize, out_wei
...
+#[inline(always)]
+pub fn forward(         further_blocks: &[Box<dyn BlockTrait>], 
+                        fb: &feature_buffer::FeatureBuffer, 
+                        pb: &mut port_buffer::PortBuffer) {
+    match further_blocks.split_first() {
+        Some((next_regressor, further_blocks)) => next_regressor.forward(further_blocks, fb, pb),
+        None => {},
+    }
+}
+
```
Comment (SkBlaz) [inline src/block_helpers.rs:183]: Perhaps worth removing these extra spaces at the end of the file

File: src/block_loss_functions.rs [hunk 3]
```diff
@@ -32,95 +34,153 @@ pub fn logistic(t: f32) -> f32 {
...
+    fn get_num_output_slots(&self) -> usize {1}   
+
+
+    fn get_num_output_values(&self, output: graph::OutputSlot) -> usize {
+        assert!(output.get_output_index() == 0);
+        1
+    }
+    
+    fn set_input_offset(&mut self, input: graph::InputSlot, offset: usize)  {
+        assert!(input.get_input_index() == 0);
```
Comment (SkBlaz) [inline src/block_loss_functions.rs:85]: Assertions are in Rust checked also in release builds, we could try without at some point (if not critical function this does not matter of course)

File: src/block_lr.rs [hunk 4]
```diff
@@ -4,110 +4,161 @@ use crate::optimizer;
...
+                    let general_gradient = myslice.get_unchecked(hashvalue.combo_index as usize);
                     let gradient = general_gradient * feature_value;
                     let update = self.optimizer_lr.calculate_update(gradient, &mut self.weights.get_unchecked_mut(feature_index).optimizer_data);
-                    self.weights.get_unchecked_mut(feature_index).weight += update;
+                    self.weights.get_unchecked_mut(feature_index).weight -= update;
                 }
             }
-            (prediction_probability, general_gradient)
+
+            return
```
Comment (SkBlaz) [inline src/block_lr.rs:134]: Do we need this return?

File: src/block_normalize.rs [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+                        update:bool) {
+        debug_assert!(self.output_offset != usize::MAX);
+        debug_assert!(self.input_offset != usize::MAX);
+        debug_assert!(self.num_inputs > 0);
+        
+        unsafe {
+            let mut sum: f32 = 0.0;
+            let mut sumsqr: f32 = 0.0;
+            let K = 1.0;
+/*          for i in 0..self.num_inputs as usize {
```
Comment (SkBlaz) [inline src/block_normalize.rs:89]: Can this be removed?

File: src/block_normalize.rs [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+// This is purely variance normalization as described in 
+// https://arxiv.org/pdf/2006.12753.pdf
+// Early results show no improvements for normalization od neural layers
+
+pub fn new_stop_block(  bg: &mut graph::BlockGraph, 
+                        mi: &model_instance::ModelInstance,
+                        input: graph::BlockPtrOutput
+                        ) -> Result<graph::BlockPtrOutput, Box<dyn Error>> {    
+    let num_inputs = bg.get_num_output_values(vec![&input]);
+    assert!(num_inputs != 0);
```
Comment (SkBlaz) [inline src/block_normalize.rs:197]: Should maybe be `debug_assert!`

File: src/block_normalize.rs [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+                    lr_buffer: Vec::new(),
+                    ffm_buffer: Vec::new(),
+                    ffm_fields_count: 0,
+        }
+    }
+
+
+    #[test]
+    fn test_simple_positive() {
+        let mut mi = model_instance::ModelInstance::new_empty().unwrap();
```
Comment (SkBlaz) [inline src/block_normalize.rs:309]: Is this commented out intentionally?

File: src/model_instance.rs [hunk 8]
```diff
@@ -90,25 +111,27 @@ impl ModelInstance {
...
             power_t: 0.5,
             ffm_power_t: 0.5,
             add_constant_feature: true,
             feature_combo_descs: Vec::new(),
             ffm_fields: Vec::new(),
             ffm_k: 0,
             ffm_bit_precision: 18,
-            ffm_separate_vectors: false, // DEPRECATED, UNUSED
             fastmath: true,
 			ffm_initialization_type: String::from("default"),
```
Comment (SkBlaz) [inline src/model_instance.rs:122]: Indent here is off

**Review by SkBlaz — CHANGES_REQUESTED** [review]
Nice work! Left some general comments.

**Comment by bbenshalom** [comment]
Please clean up all occurrences of commented out code that was added

File: src/block_helpers.rs (hunk 2 above)
Comment (adischw) [reply src/block_helpers.rs:183]: done

File: src/block_lr.rs (hunk 4 above)
Comment (adischw) [reply src/block_lr.rs:134]: removed

File: src/block_normalize.rs (hunk 5 above)
Comment (adischw) [reply src/block_normalize.rs:89]: done

File: src/block_normalize.rs (hunk 6 above)
Comment (adischw) [reply src/block_normalize.rs:197]: done

File: src/block_normalize.rs (hunk 7 above)
Comment (adischw) [reply src/block_normalize.rs:309]: it was commented out in the deep and deep-no-logs branches so i left it the same

File: src/model_instance.rs (hunk 8 above)
Comment (adischw) [reply src/model_instance.rs:122]: not sure why its showing like this but in the code indent seems fine

File: src/main.rs [hunk 9]
```diff
@@ -4,6 +4,9 @@
 #![allow(unused_mut)]
 #![allow(non_snake_case)]
 #![allow(redundant_semicolons)]
+//#[global_allocator]
+//static GLOBAL: jemallocator::Jemalloc = jemallocator::Jemalloc;
```
Comment (SkBlaz) [inline src/main.rs:8]: Let's remove these two commented-out lines too, currently pinning the allocator is not all that relevant

File: src/model_instance.rs (hunk 8 above)
Comment (SkBlaz) [reply src/model_instance.rs:122]: Ok, let's address this systematically in a separate PR then.

File: Cargo.toml [hunk 10]
```diff
@@ -11,46 +11,43 @@ edition = "2018"
...
-# We'll use cloudflare's zlib as it is the fastest game in town
-#flate2 = "1.0" #minz library
-#flate2 = { version = "1.0", features = ["zlib"], default-features = false }
+rand = "0.8.5"
+rand_distr = "0.4.3"
+rand_xoshiro = "0.6.0"
 flate2 = { version = "1.0", features = ["cloudflare_zlib"], default-features = false }
 shellwords = "1.1.0"
+blas = "0.22"
+#jemallocator = "0.5.0"
```
Comment (SkBlaz) [inline Cargo.toml:29]: Let's remove this commented-out line too

**Review by SkBlaz — APPROVED** [review]
Thanks for addressing the comments; added a few smaller ones, but approving as such (some things are to be addressed in separate PRs to separate the changes better)

File: src/main.rs (hunk 9 above)
Comment (adischw) [reply src/main.rs]: done

**Pull request merged**